dvc/stage/utils.py

```diff
@@ -251,3 +262,98 @@ def split_params_deps(stage):
     from ..dependency import ParamsDependency
 
     return lsplit(rpartial(isinstance, ParamsDependency), stage.deps)
+
+
+def _is_valid_name(name: str):
+    from . import INVALID_STAGENAME_CHARS
+
+    return not INVALID_STAGENAME_CHARS & set(name)
+
+
+def _get_file_path(kwargs):
+    """Determine file path from the first output name.
+
+    Used in creating .dvc files.
+    """
+    from dvc.dvcfile import DVC_FILE, DVC_FILE_SUFFIX
+
+    out = first(
+        concat(
+            kwargs.get("outs", []),
+            kwargs.get("outs_no_cache", []),
+            kwargs.get("metrics", []),
+            kwargs.get("metrics_no_cache", []),
+            kwargs.get("plots", []),
+            kwargs.get("plots_no_cache", []),
+            kwargs.get("outs_persist", []),
+            kwargs.get("outs_persist_no_cache", []),
+            kwargs.get("checkpoints", []),
+            without([kwargs.get("live", None)], None),
+        )
+    )
+
+    return (
+        os.path.basename(os.path.normpath(out)) + DVC_FILE_SUFFIX
+        if out
+        else DVC_FILE
+    )
+
+
+def _check_stage_exists(dvcfile, stage):
+    from dvc.stage import PipelineStage
+    from dvc.stage.exceptions import (
+        DuplicateStageName,
+        StageFileAlreadyExistsError,
+    )
+
+    if not dvcfile.exists():
+        return
+
+    hint = "Use '--force' to overwrite."
+    if stage.__class__ != PipelineStage:
+        raise StageFileAlreadyExistsError(
+            f"'{stage.relpath}' already exists. {hint}"
+        )
+    elif stage.name and stage.name in dvcfile.stages:
+        raise DuplicateStageName(
+            f"Stage '{stage.name}' already exists in '{stage.relpath}'. {hint}"
+        )
+
+
+def create_stage_from_cli(
+    repo: "Repo", single_stage: bool = False, fname: str = None, **kwargs: Any
+) -> "Stage":
+    from dvc.dvcfile import PIPELINE_FILE, Dvcfile
+    from dvc.exceptions import OutputDuplicationError
+
+    from . import PipelineStage, Stage, create_stage, restore_meta
+
+    if single_stage:
+        kwargs.pop("name", None)
+        stage_cls = Stage
+        path = fname or _get_file_path(kwargs)
+    else:
+        stage_name = kwargs.get("name", None)
+        path = PIPELINE_FILE
+        stage_cls = PipelineStage
+        if not (stage_name and _is_valid_name(stage_name)):
+            raise InvalidStageName
+
+    params = chunk_dict(parse_params(kwargs.pop("params", [])))
+    stage = create_stage(
+        stage_cls, repo=repo, path=path, params=params, **kwargs
+    )
+    restore_meta(stage)
+
+    try:
+        if kwargs.get("force", True):
+            with suppress(ValueError):
+                repo.stages.remove(stage)
```

Comment:
Thinking of keeping `stages` in a dictionary as `repo.stage_d` and maintain compatibility by returning `stage_d.values()` as a `repo.stages`.